test/scripts/e2e_subs/min_balance.py

```diff
@@ -0,0 +1,184 @@
+#!/usr/bin/env python
+
+from datetime import datetime
+from pathlib import PurePath
+import sys
+
+import algosdk.future.transaction as txn
+
+from goal import Goal, AtomicABI
+
+CONSENSUS_MIN_BALANCE = 100_000
+ASSET_MIN_BALANCE = 100_000
+APP_MIN_BALANCE = 100_000
+OPTIN_MIN_BALANCE = 100_000
+# app schemas:
+APP_KV_MIN_BALANCE = 25_000
+APP_INTS_MIN_BALANCE = 3_500
+APP_BYTES_MIN_BALANCE = 25_000
+
+EXTRA_PAGE_MIN_BALANCE = (
+    APP_MIN_BALANCE  # per userBalance.go::MinBalance() as of Dec 2021
+)
+
+
+# Set INTERACTIVE True if you want to run a remote debugger interactively on the given PORT
+INTERACTIVE, DEBUGPORT = False, 4312
+
+
+def initialize_debugger():
+    import multiprocessing
+
+    if multiprocessing.current_process().pid > 1:
+        import debugpy
+
+        debugpy.listen(("0.0.0.0", DEBUGPORT))
+        print("Debugger is ready to be attached, press F5", flush=True)
+        debugpy.wait_for_client()
+        print("Visual Studio Code debugger is now attached", flush=True)
+
+
+if INTERACTIVE:
+    initialize_debugger()
+
+
+def get_pysdk_min_balance(goal, account):
+    return goal.algod.account_info(account)["min-balance"]
+
+
+def get_teal_min_balance(abi, account):
+    # can't execute an abi object twice so must clone it before each execution:
+    abi = abi.clone(caller_acct=account)
+    sender_pymt = txn.PaymentTxn(account, abi.get_suggested_params(), account, 10_000)
+    sender_tx_sig = abi.get_txn_with_signer(sender_pymt)
+
+    # TODO: abi.sender_min_balance(sender_tx_sig)
+    return abi.execute_singleton("sender_min_balance", method_args=[sender_tx_sig])
+
+
+def assert_min_balance(abi_or_goal, account, expected_min_balance, skip_abi=False):
```

Comment:
Query the `min_balance` in two ways, and compare the results and also to what we think they should be. The two ways are:
1. Via **algod REST** / **Python SDK**
2. Via **AtomicABI** to query a TEAL app that uses the `min_balance` op code

Reply:
I'm in favor of using python instead of an unholy combination or shell/goal/jq/curl/grep etc, so tests are clearer.  But I think the ABI stuff is overboard here. A very, very simple teal app that returns min_balance could be instantiated and called using just the sdk basics that were already here.

As a general rule, if we just use python as a more convenient way to write good tests, and stay away from testing the most recent features (ABI in this case) we will get better tests, but we can avoid difficulties with pinning the version.  Just use old tried and true SDK functions.

Reply:
👍